support various pod traffic encapsulation mode

suwang48404 opened the pull request:
Design doc
https://docs.google.com/document/d/1X-UXz38XcFfLkjWE_5ke_g-WYRwsTaA04182OGDtO4g/edit#heading=h.8haslsej50qf

suwang48404 commented:
/test-e2e
/test-conformance
/test-networkpolicy

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
/test-e2e

antoninbas commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
@jianjuns @antoninbas @tnqn 
Hi guys, do u mind have a look at this PR? The design doc is also included.

thx, Su

tnqn commented on pkg/agent/types/node_config_win.go:
```diff
@@ -0,0 +1,81 @@
+		"ServiceCIDR: %s\nServieRT: %s", n.Name, n.PodCIDR, n.NodeIPAddr, n.PodEncapMode, n.GatewayConfig,
+		n.ServiceCIDR, n.ServiceRtTable)
+}
+*/
```
If windows is not going to work anyway for now, maybe don't add it to save some review efforts.

suwang48404 replied:
node_config_win.go is removed,

tnqn commented on build/yamls/antrea.yml:
```diff
@@ -264,6 +264,13 @@ data:
+    # - noEncap
+    # - hybrid
+    # - noEncapMasq
+    #podTrafficEncapMode: encap
```
Maybe add more details here about how each mode works, as this is user facing.
BTW, does `noEncapMasq` mean `passthrough` @jianjuns tried to add, I feel `noEncapMasq` might be a little confusing about whether it's going to masquerade or not and what's the masquerade behavior of `encap` and `noEncap`, if it indeed means `passthrough`, the later sounds more clear to me.

suwang48404 replied:
yes, we can rename to passthrough/bump-in-wire,  if we are on the same page

suwang48404 replied:
@jianjuns @antoninbas 
what do u think rename NoEncapMasq to 
1) passthrough or
2) BumpInWire

Thx, Su

jianjuns replied:
I am not really good at naming, but was thinking about passthrough or underlayNetwork. @antoninbas @abhiraut @salv-orlando might have some better ideas.

jianjuns replied:
Today @antoninbas, @abhiraut, and I chatted about this a little. Our thinking is that as we still need to close the design with public cloud (I thought we should be able to close it faster and then need to have the mode ready, but seems this is not the case), it is not very clear how this mode for public clouds will be exactly like. In this case, should we make it a separate PR and merge after we close the design? Otherwise seems hard to make sure we implement something really useful, or to discuss the naming. Let me know what you think.

suwang48404 replied:
yes, noEncapMasq assumes there is underlaying CNI network already working (as in managed AKS, EKS), and Antrea is only do ingress/egress functionaries

suwang48404 replied:
regarding to how this change relates to native public cloud support, please see design doc. It is mentioned a bit.
At this point, I am looking to use NoEncapMasq in conjunction with AKS and EKS CNI. That is AKS/EKS provide connectivity, Antrea provides policy

suwang48404 commented on cmd/antrea-agent/agent.go:
```diff
@@ -16,6 +16,7 @@ package main
 
 import (
 	"fmt"
+	"github.com/vmware-tanzu/antrea/pkg/agent/route"
```
done

suwang48404 commented on cmd/antrea-agent/config.go:
```diff
@@ -65,4 +65,11 @@ type AgentConfig struct {
 	// Defaults to false.
 	EnableIPSecTunnel bool `yaml:"enableIPSecTunnel,omitempty"`
+	// Determines how pod traffic is encapsulated. It has the following options
+	// Encap: inboud(stateful) traffic always encapsulated. outbound masqueraded. This is default
```
done

suwang48404 commented on cmd/antrea-agent/config.go:
```diff
@@ -65,4 +65,11 @@ type AgentConfig struct {
+	// Hybrid: inbound noEncap if worker nodes on same subnet, otherwise encap. Masq for outbound.
+	// NoEncapMasq: no encap and masq.
+	PodTrafficEncapMode string `yaml:"podTrafficEncapMode,omitempty"`
+	TestArg             string `yaml:"testArg,omitempty"`
```
done

suwang48404 commented on cmd/antrea-agent/config.go:
```diff
@@ -65,4 +65,11 @@ type AgentConfig struct {
+	// Determines how pod traffic is encapsulated. It has the following options
+	// Encap: inboud(stateful) traffic always encapsulated. outbound masqueraded. This is default
+	// NoEncap: no encap for inbound, but masq for outbound
+	// Hybrid: inbound noEncap if worker nodes on same subnet, otherwise encap. Masq for outbound.
```
done

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -119,6 +146,58 @@ func (c *Client) SetupRules() error {
 		}
 	}
+
+	signals.ExitFns.Add(func() error {
```
considering:

1. k8s pods by nature are ephemeral. Application has the ability to re-establish service. From application's perspective, this is no different than a service backing pod goes down, the service connection is re-established to another service backing pod.
2. place nice to next CNI getting installed, it could be an new version of Antrea or a CNI from different vendor which are unaware (old) Antrea and iptable rules which can break their operation.

antoninbas replied:
Proper deletion / cleanup is a separate concern IMO. The dataplane should not break IMO when the antrea agent is stopped (as part of a planned upgrade for example), and I would say this concern takes precedence.

There is a need to handle cleanup with a separate mechanism any way, since there is a always a possibility that Antrea does not exit gracefully when deleted.

suwang48404 replied:
true, cleanup is separate issue. Will comment out.

suwang48404 replied:
I gave some thoughts on this, and I just don't feel comfortable to have lingering route table entries or iptable entries, because next CNI may be
1. an non-antrea CNI or an Antrea upgrade that is unaware of the route table and iptable entries being added.
2. the same Antrea CNI but operates on different mode that is unaware of the route table and iptable entries

Before encap mode support, route table and iptable more or less stay static between each agent start, so there is little motivation to clear up them. Now with encap mode supports, iptable and route table are different when encap mode are different between each restart. So it is prudent to cleanup. 

Hope this make sense,

jianjuns replied:
For non-antrea CNI, we plan to have a separate cleanup mechanism after Antrea uninstall. @antoninbas has a proposal (#307) on this.
For other upgrade/config change case, Antrea should handle the cleanup after restart, or before it can handle that we have to require the separate cleanup step.
Agreed with what others said, we should first make sure no traffic outage when Agent is down/restarted.

suwang48404 replied:
My thinking is
between Antrea restart:
1. if Antrea upgrade, the new version may have changed, the route table and iptables entries it modifies may also change, 
2. if Antrea change to another encapsulation mode, the new incarnation does not know what iptable and route table entries has been changed in previous incarnation

suwang48404 replied:
I don't believe this change is high risk, it makes changing mode on the fly possible

suwang48404 replied:
to continue my thoughts,
fundamentally I think it is important for Antrea to play nice with next CNI installed, be that newer version of Antrea, Antrea running on different mode, or another CNI as suggested by  https://github.com/vmware-tanzu/antrea/pull/307

From workflow perspective, keep in mind, with new noEncap/hybrid mode, iptables and route entries are different from those of encap mode.  this change allows Antrea simply to be reapplied without any other work. Without it, at least for encapsulation mode change (or even further upgrade), user may need to manually or an out-of-band script (may behttps://github.com/vmware-tanzu/antrea/pull/307) to remove iptables and route table entries from last Antrea installation before reapply with new encap mode.

From traffic outage perspective,
-- if upgrade, I am guessing both ovs/agent will be restarted, so traffic will be interrupted regardless
-- if agent crashed, today we capture only sigterm, sigint, so crash should not cause entries being wiped out.
-- and once again, k8s is build-in fault tolerant, From application level, all that matters is if there is service outage, and this change will not cause service outage.

Hope this make sense, Su

tnqn replied:
I see normally upgrading/changing mode relies on next round to clean up the previous configuration, instead of previous round cleaning up itself, as it can be guaranteed previous round is terminated gracefully, a round might crash in the middle of setting iptables and routes, so it has to do the clean up anyway (and antrea already does such cleanup at startup). 
A new round doesn't need to know what mode or configuration was used in previous rounds, it should always clean up all iptables, routes managed by antrea and set up desired ones, of course in-place update is better. For example, all iptables rules managed by antrea should be in its own chains, so we can delete previous ones and install new ones (iptables-restore can make it easier).

tnqn replied:
A similar case is kube-proxy, it won't do any cleanup when exiting, but cleans up previous iptables rules and sets up new ones after starting regardless of how its implementation is changed.

suwang48404 replied:
it is much harder to clean up at restart, new incarnation do not know what iptable chains is used by previous incarnation. Iptable-save/restore would help, but there are not there yet today. Same goes for routing tables, given that we can have multiple route tables. 

but it is clear we prefer cleaning up on restart, will change accordingly. And open another PR to address clean up at restart.

suwang48404 replied:
done. route and iptables are reconciled at agent starting up,

antoninbas commented on Makefile:
```diff
@@ -65,6 +69,20 @@ docker-test-unit: $(DOCKER_CACHE)
 	@chmod -R 0755 $<
 
+.PHONY: docker-test-integration
+docker-test-integration:
```
This should be part of a separate PR, this PR is large enough as it is.

suwang48404 replied:
I used docker-test-integration for hew route_test.go which changes route tables hence only implemented when test is running is container. Only it is Ok this be part of PR, but separate commit.

antoninbas commented on docs/kind.md:
```diff
@@ -56,6 +56,14 @@ antrea-agent-zsztq                   2/2     Running   0          8m56s
+### Short Cut
+ Alternatively to create a two worker node cluster with antrea installed, do
+
+  ./hack/kind-setup.sh create CLUSTER_NAME
```
I don't think it's great for users to have to use a modified version of kind (or in this case a script that wroas the kind binary), so I wouldn't advertise this script like this here, and probably move it to `ci/kind/`. While I appreciate the fact that your changes are useful for CI, for a user who just wants to deploy Antrea to a Kind cluster to quickly experiment with network policy support, different networking modes are irrelevant IMO.

suwang48404 replied:
ok will move to ci/kind

antoninbas commented on hack/kind-setup.sh:
```diff
@@ -0,0 +1,334 @@
+  --antrea-cni: specifies use antrea or default cni in kind cluster, default is $ANTREA_CNI
+  --num-workers: specifies number of worker nodes in kind cluster, default is $NUM_WORKERS
+  --images: specifies images loaded to kind cluster, default is $IMAGES
+  --subnetnets: sp0ecifies subnets and bridge networks used worker nodes, default is empty all worker
```
typos?

antoninbas commented on hack/kind-setup.sh:
```diff
@@ -0,0 +1,334 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is required for Antrea to work properly in a Kind cluster on Linux. It takes care of
```
This is my original comment, but your script does a lot of things now.

suwang48404 replied:
will add more comments

antoninbas commented on hack/test-e2e-kind.sh:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env bash
```
suggest moving this to `ci/kind/`

suwang48404 replied:
will do

antoninbas commented on pkg/agent/types/node_config_win.go:
```diff
@@ -0,0 +1,81 @@
+// Copyright 2019 Antrea Authors
```
let's not include this in this PR

suwang48404 replied:
this is required to compile windows antctl. will change the header though

antoninbas replied:
It seems that a better solution would be to refactor these type definitions and move them to another OS-agnostic package, instead of duplicating them. What is Windows-specific in this file?

suwang48404 replied:
I agree, that would be idea. Perhaps another PR?

suwang48404 replied:
node_config_win is removed after some code re-structure.

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -16,6 +16,10 @@ package iptables
 	"fmt"
+	"github.com/vmware-tanzu/antrea/pkg/agent/openflow"
+	"github.com/vmware-tanzu/antrea/pkg/agent/types"
+	"github.com/vmware-tanzu/antrea/pkg/signals"
```
done

antoninbas commented on test/e2e/framework.go:
```diff
@@ -17,6 +17,7 @@ package e2e
 import (
 	"bytes"
 	"fmt"
+	"github.com/vmware-tanzu/antrea/pkg/agent/types"
```
@suwang48404 at the moment tests can be run from macOS (by connecting to an arbitrary K8s cluster running Antrea). I hope this import does not break that?

suwang48404 replied:
good catch, that would break Mac os test. thx

suwang48404 replied:
reference to netlink in pkt/agent/types module is removed, so it can safely called on Mac OS

jianjuns commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -32,3 +32,10 @@
 # be set to the same value as the one specified by --service-cluster-ip-range for kube-apiserver.
 #serviceCIDR: 10.96.0.0/12
+
+# Determines how pod traffic is encapsulated. It has the following options
```
We generally capitalize the first letter of K8s resource type, like Pod, Node, NetworkPolicy, etc..

jianjuns commented on pkg/agent/agent.go:
```diff
@@ -109,6 +117,11 @@ func (i *Initializer) GetIPSecPSK() string {
 }
 
+// GetNodeConfig returns iptables client
+func (i *Initializer) GeIPTablesClient() *iptables.Client {
```
Typo in GeIPTablesClient

jianjuns commented on pkg/agent/agent.go:
```diff
@@ -109,6 +117,11 @@ func (i *Initializer) GetIPSecPSK() string {
 	return i.ipsecPSK
 }
 
+// GetNodeConfig returns iptables client
```
GetNodeConfig -> GetIPTablesClient

jianjuns commented on cmd/antrea-agent/agent.go:
```diff
@@ -103,7 +106,9 @@ func run(o *Options) error {
-		agentInitializer.GetIPSecPSK())
+		agentInitializer.GetIPSecPSK(),
+		routeClient,
+		agentInitializer.GeIPTablesClient())
```
Humm. would you consider to move the IPTablesClient creation to this file just like route.NewClient() in line 72?

jianjuns commented on pkg/agent/types/pod_encap_mode.go:
```diff
@@ -0,0 +1,73 @@
+	"strings"
+)
+
+type PodEncapMode int
```
I feel TrafficEncapMode is better. It can be Node -> Pod traffic, and even Node -> GW traffic as well.

suwang48404 replied:
done

jianjuns commented on cmd/antrea-agent/options.go:
```diff
@@ -127,8 +132,15 @@ func (o *Options) setDefaults() {
-		if o.config.TunnelType == ovsconfig.VXLANTunnel {
+		ok, encapMode := types.GetPodEncapModeFromStr(o.config.PodTrafficEncapMode)
+		if ok && !encapMode.SupportsEncap() {
+			o.config.DefaultMTU = defaultMTU
```
We should use defaultMTU only for encap, but not hybrid?

suwang48404 replied:
supportEncap include encap and hybrid mode

jianjuns replied:
Got it.

jianjuns commented on pkg/agent/agent.go:
```diff
@@ -342,11 +353,12 @@ func (i *Initializer) setupGatewayInterface() error {
 	// restarts.
 	klog.V(4).Infof("Setting gateway interface %s MTU to %d", i.hostGateway, i.mtu)
-	i.ovsBridgeClient.SetInterfaceMTU(i.hostGateway, i.mtu)
+	_ = i.ovsBridgeClient.SetInterfaceMTU(i.hostGateway, i.mtu)
```
Any reason to change this?

suwang48404 replied:
by convention, go expect u to handle error returned by function. If choose not to, do so explicitly.

jianjuns commented on pkg/agent/agent.go:
```diff
@@ -410,11 +427,25 @@ func (i *Initializer) setupGatewayInterface() error {
 func (i *Initializer) setupDefaultTunnelInterface(tunnelPortName string) error {
 	tunnelIface, portExists := i.ifaceStore.GetInterface(tunnelPortName)
+
+	if !i.nodeConfig.PodEncapMode.SupportsEncap() {
```
Like to check @antoninbas if he has a different idea on how to cleanup.

jianjuns commented on pkg/agent/agent.go:
```diff
@@ -454,11 +485,21 @@ func (i *Initializer) initNodeLocalConfig() error {
+		return fmt.Errorf("initNodeLocalConfig %v", err)
+	}
+
+	servRtTable := &types.ServiceRtTableConfig{Idx: 254, Name: "main"}
```
Could we move line 488-497 to package route, as it needs specific knowledge like Idx/Name?

suwang48404 replied:
done.

jianjuns commented on pkg/agent/agent.go:
```diff
@@ -454,11 +485,21 @@ func (i *Initializer) initNodeLocalConfig() error {
+		servRtTable.Idx = 300
+		servRtTable.Name = "Antrea-service"
+	}
+	i.nodeConfig = &types.NodeConfig{Name: nodeName, PodCIDR: localSubnet, NodeIPAddr: localAddr, NodeDefaultDev: dev.Attrs().Name, PodEncapMode: i.podEncapMode, ServiceCIDR: i.serviceCIDR, ServiceRtTable: servRtTable}
```
I need to look into these, but wonder: 1) do we put all them in NodeConfig, esp. if they are not shared by other NodeConfig consumers. Now NodeConfig just has a small set of Node info. 2) do we need to initialize and also save (all) them in agentInitializer. For example, are some of them used by only routeClient?

suwang48404 replied:
routable is now moved out of NodeConfig

suwang48404 commented:
/test-all

suwang48404 commented:
@jianjuns @antoninbas @tnqn 
thx for spending time reviewing this large PR. I hope I have addressed most if not all of ur comments. I wonder if I can check is in with ur approval?
thx,

antoninbas commented on test/e2e/connectivity_test.go:
```diff
@@ -104,7 +104,7 @@ func createPodsOnDifferentNodes(t *testing.T, data *TestData, numPods int) (podN
 
 func (data *TestData) testPodConnectivityDifferentNodes(t *testing.T) {
-	numPods := 2 // can be increased
+	numPods := clusterInfo.numNodes
```
why? If we run on a large cluster, this can significantly increase the duration of the test, given that Pods are created sequentially.

suwang48404 replied:
I only want to do this in hybrid mode where test both encap and noencap traffic. Will change

antoninbas commented on ci/kind/test-e2e-kind.sh:
```diff
@@ -0,0 +1,59 @@
+docker pull nginx
+
+echo "======== Test encap mode =========="
+run_test encap '--images "busybox nginx antrea/antrea-ubuntu:latest"'
```
maybe a stupid question, but why do we bother loading busybox & nginx into the Kind cluster, instead of letting the Nodes download it? Speed?

maybe define a bash variable instead of repeating the image list 3 times

suwang48404 replied:
yes for speed, instead of loading images from internet on each node.  may not help much in the end,

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -79,6 +101,12 @@ type rule struct {
 	comment string
 }
 
+// Initialize sets up internal variables
```
ok will give it audit

suwang48404 replied:
I wonder if go lint would have helped in this case?

suwang48404 replied:
done.

suwang48404 commented on build/yamls/antrea-ipsec.yml:
```diff
@@ -264,6 +264,13 @@ data:
+    # encap(default): inter-node Pod traffic is always encapsulated and Pod to outbound traffic is masqueraded.
+    # noEncap: inter-node Pod traffic is not encapsulated, but Pod to outbound traffic is masqueraded
+    # hybrid: noEncap if worker Nodes on same subnet, otherwise encap.
+    # noEncapMasq: Pod traffic is not encapsulated nor masqueraded.
```
@jianjuns @tnqn 

any input on name change for noEncapMasq ==>
1) passthrough
2)bumpInWire

thx

antoninbas replied:
BumpInTheWire is probably more precise in a way. But for people not familiar with the networking term, "passthrough" is probably clearer. I would vote for passthrough, since the term will need to be documented any way, but let's hear what @jianjuns has to say.

suwang48404 commented on build/images/scripts/start_ovs_netdev:
```diff
@@ -74,8 +74,11 @@ function quit {
-
 start_ovs
+
+if [[ "$#" -ge 1 ]] && [[ "$1" == "--no-monitor" ]]; then
```
good idea

suwang48404 replied:
moved commands to a separate bash : test-integration

suwang48404 commented:
@antoninbas 

thx for reviewing. couples of responses to ur questions

1 regarding to noEncapMasq, it is just a place holder right now, It need to integrated with under network, for instance, AKS, EKS, GPC etc. Thus there is no test bed for them
2 regarding to kind e2e length, I agree we should do something about it. I have something working based on this PR that uses a single testbed run across all encap mode, and only run selective redundant tests on each encap mode. Once this PR is checked in, I will PR that change.

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,333 @@
+kind: Cluster
+apiVersion: kind.sigs.k8s.io/v1alpha3
+networking:
+  disableDefaultCNI: true
```
we can do that. But I choose not to install any CNI at all if ANTREA_CNI is false. Because I do not know the consequence of install Antrea on top Kubelet default CNI, and wants to keep it safe.

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,333 @@
+  --antrea-cni: specifies use antrea or default cni in kind cluster, default is $ANTREA_CNI
+  --num-workers: specifies number of worker nodes in kind cluster, default is $NUM_WORKERS
+  --images: specifies images loaded to kind cluster, default is $IMAGES
+  --subnetnets: specifies subnets and bridge networks used worker nodes, default is empty all worker
```
reworded comment for --subnets

suwang48404 commented on cmd/antrea-agent/config.go:
```diff
@@ -65,4 +65,10 @@ type AgentConfig struct {
 	// through an environment variable: ANTREA_IPSEC_PSK.
 	// Defaults to false.
 	EnableIPSecTunnel bool `yaml:"enableIPSecTunnel,omitempty"`
+	// Determines how pod traffic is encapsulated. It has the following options
```
done

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,293 @@
+
+var (
+	// ServiceRtTable contains Antrea service route table information.
+	ServiceRtTable = &serviceRtTableConfig{Idx: 254, Name: "main"}
```
Is ServiceRtTable used outside the route package? If not make it internal - serviceRtTable.

suwang48404 replied:
ServiceRtTable is also used in integration test. Otherwise it can be made local

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,293 @@
+		} else {
+			routes = append(routes, &netlink.Route{
+				Dst:       peerPodCIDR,
+				Flags:     int(netlink.FLAG_ONLINK),
```
Question - why we need the ONLINK flag and specify the dev? For hybrid mode, I guess we need not this. So, maybe you mean for noEncap mode, you want to force the packet to be sent to the destination Node IP, assuming source Node can resolve the destination Node's MAC, even it is in a different subnet?
Do you really see such scenarios (Nodes can reach each other even they are not in the same subnet)? I was thinking if we really want to support NoEncap mode (besides hybrid. You know I do not really suggest to add NoEncap which is not very useful), we might assume someone or some code configured underlay router to route Pod traffic across different Node subnets, and in this case we should not add routes in the source Node. What you think?

suwang48404 replied:
yes, the onlink flag is to be safe to noEncap mode where peer IP is not on the same subnet but is on the same physical network.

suwang48404 replied:
I hope we keep to keep NoEncap mode as it provide additional flexibility to our user.

jianjuns replied:
K8s Nodes are not really that ephemeral - they are VMs not Pods, and needs to install K8s and join the Cluster, typically add/removed manually (I mean manually triggered).
Even you believe manual route configuration is not common, do you think there can another controller to automate router configuration at Node creation/removal? How about your DHCP idea?

But do you see real deployments of using multiple subnets in a single physical L2/VLAN? I guess it is real.

Again, as I said earlier, in general I do not think NoEncap is very useful in either of the two cases.

jianjuns replied:
Of course, one possible route automation is by BGP.

suwang48404 replied:
1. regarding to how worker node IP can be changed. My experience is that user does not need to remove and add node to k8s cluster. An power off/on of worker node is suffice to change its node ip, assuming kubelet is started with --node-ip DHCLIENT_IP. 

2.regarding to the usefulness to NoEncap mode, my thinking is that we don't know what users deployment case looks like, why not give that option? And we know in circumstances that multiple IP subnets may be accessible on the same L2 network. I personally do not believe that hybrid covers all potential use cases,  that is reason it make sense to me that we give the user to option to configure NoEncap and Encap in addition to hybrid.


But I diverse, the question is that in NoEncap mode, if peerNode is not on the same subnet, should peerPodCIDR nexthop be peerNode(IP) with onlink (this works if all worker nodes are reachable via L2); or nexthop be default gw (this works if  underlay router knows how to route pod CIDR traffic)?

jianjuns replied:
Yes. I agree with your last summary. So, we have three options:
1. Do not support NoEncap at all, if no real use case. It will also be hard to test.
2. Define it as all Nodes in the same physical L2, even they are not in the same subnet - your current implementation.
3. Define it as Node route table and underlay router will handle the routing, which can have two cases: 1) user manually configure routes; 2) there is another solution automatically configures routes or uses BGP to advertise routes. Last how if the Pod IPs are allocated by DHCP - it should be treated as noEncap or passthrough?

I do not quite like #2 because I assume most people will think about #3 when talking about noEncap, esp. with other solutions like Calico.
If you like, we can have more offline discussion.

suwang48404 replied:
renamed noEncap to noEncapSameL2 as discussed offline

suwang48404 commented on pkg/agent/types/traffic_encap_mode.go:
```diff
@@ -0,0 +1,73 @@
+const (
+	TrafficEncapModeEncap TrafficEncapMode = iota
+	TrafficEncapModeNoEncap
+	TrafficEncapModePassThrough
```
my goland editor seem to think they are two words. see also, https://www.dictionary.com/browse/passthrough

suwang48404 commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -32,3 +32,10 @@
+# noEncap: inter-node Pod traffic is not encapsulated, but Pod to outbound traffic is masqueraded
+# hybrid: noEncap if worker Nodes on same subnet, otherwise encap.
+# passThrough: antrea-agent provides network services at pod ingress/egress. Network connectivity is delegated to underlying network.
+#trafficEncapMode: encap
```
I don't see them. From which line?

suwang48404 replied:
done

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -260,10 +286,23 @@ func (c *client) initialize() error {
+			return fmt.Errorf("failed to install l2 forward in port out flow: %v", err)
+		}
+	}
+	if c.nodeConfig.TrafficEncapMode == types.TrafficEncapModePassThrough {
```
If u want, I can remove all reference to passthrough mode. But I'd hate to having to come back again redo the same work when supporting EKS/AKS

suwang48404 replied:
done. removed all reference to passthrough

suwang48404 commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -469,21 +465,12 @@ func (c *Controller) addNodeRoute(nodeName string, node *v1.Node) error {
-	if err = netlink.RouteReplace(route); err != nil {
-		return fmt.Errorf("failed to install route to Node %s with netlink: %v", nodeName, err)
+	routes, err := c.routeClient.AddPeerCIDRRoute(peerPodCIDR, c.gatewayLink.Attrs().Index, peerNodeIP, peerGatewayIP)
+	c.installedNodes.Store(nodeName, routes)
```
there maybe multiple routes being added for service and main route table.

but this perhaps is not correct logic, err should be return if any routes cannot be added.

suwang48404 replied:
done. Reworked the logic to AddPeerCIDRRoute that when encounter a route add error, it deletes any route already added, and return err with empty route list.

suwang48404 commented on pkg/agent/agent.go:
```diff
@@ -454,11 +485,16 @@ func (i *Initializer) initNodeLocalConfig() error {
+		return fmt.Errorf("initNodeLocalConfig %v", err)
+	}
+
+	i.nodeConfig = &types.NodeConfig{Name: nodeName, PodCIDR: localSubnet, NodeIPAddr: localAddr, NodeDefaultDev: dev.Attrs().Name, TrafficEncapMode: i.trafficEncapMode, ServiceCIDR: i.serviceCIDR}
```
I thought nodeConfig is global structure that contains all global parameters pertain to agents.  Would rename it to NetworkConfig better in ur view?

jianjuns replied:
The problem is NodeConfig now includes some Node's local network configurations, which ARE NOT from user config parameters from the ConfigMap, but some from K8s Node resources, some further retrieved from host networking. They are different global parameters like encapMode, enableIPSec, IPSecPSK, etc. which come from users and are share by all Nodes.

jianjuns replied:
For you last comments, encapMode/IPSec/tunnelType/ServiceCIDR, etc. seems need to be shared across Nodes.
And another my point is their source (which is from user config) is different the current NodeConfig (which is from K8s Node resource or host local networking state).

suwang48404 replied:
Ok. I think I see what u mean, Will move TrafficEncapMode, ServiceCIDR outside of NodeConfig. Make sense to me.

suwang48404 replied:
move encapMode out of NodeConfig, put it into a separate pkg/agent/config module.
move serviceCIDR out of NodeConfig, and pass it directly to consumers.

suwang48404 commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -32,3 +32,10 @@
+# encap(default): inter-node Pod traffic is always encapsulated and Pod to outbound traffic is masqueraded.
+# noEncap: inter-node Pod traffic is not encapsulated, but Pod to outbound traffic is masqueraded
+# hybrid: noEncap if worker Nodes on same subnet, otherwise encap.
+# passThrough: antrea-agent provides network services at pod ingress/egress. Network connectivity is delegated to underlying network.
```
will do

suwang48404 replied:
done

suwang48404 commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -103,10 +107,13 @@ func NewNodeRouteController(
 		interfaceStore:   interfaceStore,
 		nodeConfig:       config,
-		gatewayLink:      link,
+		gatewayLink:      util.GetNetLink(config.GatewayConfig.Link),
```
will do

jianjuns replied:
Yeah. Just that my first thought is you mean the default gateway of the Node/host when I saw the field. No strong opinion.

suwang48404 replied:
I did not change the name in this PR. I guess I will keep it as is

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -101,6 +130,15 @@ func (c *Client) SetupRules() error {
 	}
 
+	if c.nodeConfig.TrafficEncapMode.SupportsNoEncap() {
+		rules = append(rules,
```
will do

suwang48404 replied:
done.

suwang48404 commented on pkg/agent/util/net_unix.go:
```diff
@@ -0,0 +1,64 @@
+// Copyright 2019 Antrea Authors
```
netlink is not supported on windows, I discovered this because antctl for windows would not compile otherwise,

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -221,12 +237,22 @@ func (c *client) InstallClusterServiceCIDRFlows(serviceNet *net.IPNet, gatewayOF
 		c.l2ForwardCalcFlow(gatewayMAC, gatewayOFPort, cookie.Default),
-		c.localProbeFlow(gatewayAddr, cookie.Default),
+	}
+	// in BumpOnWire mode, podCIDR is not known, therefore no gwIP and related flows
```
wil do

suwang48404 replied:
done. removed all reference on passthrough

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -118,6 +156,126 @@ func (c *Client) SetupRules() error {
+	if !c.nodeConfig.TrafficEncapMode.NeedEncap(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+		ruleSpec := []string{
+			"-i", c.hostGateway, "!", "-o", c.hostGateway,
+			"-d", peerPodCIDR.String(), "-s", c.nodeConfig.PodCIDR.String(), "-j", AcceptTarget,
```
that will be the best if agent knows cluster CIDR, not sure what that info comes from though?

jianjuns replied:
One way is to require user to provide it, like ServiceCIDR. @tnqn: do you know any better way?

wenyingd replied:
Is it necessary to limit the source IP must be local Pod CIDR (`"-s", c.nodeConfig.PodCIDR.String()`)? Will it block external Service traffic, like NodePort?

suwang48404 replied:
we can perhaps further limit traffic with -s podCIDR, but it may not be required because anything comes out of gw0 must come from local pod, 

traffic to service IP will not has dstIP in peerPodCIDR.

wenyingd replied:
I guess I might not present my question clearly. Let's think about this senario, an external address is accessing a NodePort Service. The source IP should be the external address, and I am wondering if the packet could be dropped by this iptables rule, for the source address is not in "c.nodeConfig.PodCIDR" or "podCIDR".

Besides,  the traffic to service IP is using Service IP as the destination, but after the kube-proxy works, the destination IP should be changed to the chosen backend Pod IP. If the Pod is on peer Node, then the destionation IP should be in peerPodCIDR.

suwang48404 replied:
1. if this is external traffic for nodePort service, this rule does not apply because traffic does not satisfy inport=gw0 and outport!=gw0

2. the reason this rule exists if so that pod-to-pod traffic does not get masqueraded, and that is all. please see comment.

tnqn replied:
> One way is to require user to provide it, like ServiceCIDR. @tnqn: do you know any better way?

AFAIK, we can only get it via a config provided by user. Even kube-proxy which needs it can only take it from config and will sacrifice some feature if not provided:
https://github.com/kubernetes/kubernetes/blob/641321c94c3e0943a57230596164e2e8f2eb2c52/cmd/kube-proxy/app/server.go#L148
https://github.com/kubernetes/kubernetes/blob/641321c94c3e0943a57230596164e2e8f2eb2c52/pkg/proxy/iptables/proxier.go#L288

suwang48404 commented:
@jianjuns 

thx for review.  I hope to get some clarity from u, 
Do u suggest
1. I remove all preferences to passthrough mode because it is incomplete feature?
2. what u suggest to do with NodeConfig structure currently it contains global configuration pertain to agent.
    --- rename to something like GlobalConfig
    --- Do away NodeConfig, and move all its content as separate agent global variables

Thx, Su

suwang48404 commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -381,13 +373,14 @@ func (c *Controller) syncNodeRoute(nodeName string) error {
+	entry, flowsAreInstalled := c.installedNodes.Load(nodeName)
+	routes := entry.([]*netlink.Route)
+	for _, r := range routes {
+		if err := netlink.RouteDel(r); err != nil {
```
DeletePeerRoute can be used here too. But DeletePeerRoute is currently called only during stale route remove stage at agent starting up where the dst is known, but nh, dev, route table is not known. 

On the other hand, netlink>routeDel is called when a known worker node is removed, and it is route information is already saved, so netlink.RouteDel is suffice.

suwang48404 commented on pkg/agent/util/net_unix.go:
```diff
@@ -0,0 +1,66 @@
+		return nil, nil, err
+	}
+	defer conn.Close()
+	localIP := conn.LocalAddr().(*net.UDPAddr).IP
```
this code does not require internet access. UDP is connectionless.  As network connect udp socket, in queries route table in how to reach internet(8.8.8.8), but doing so it finds out what is the local ip and local IP used for that udp socket. and there is no packet sent out or expected reply back.

Of course the code will be problematic is default route and route to 8.8.8.8 used different outbound IP and dev ...

Hope this helps, Su

suwang48404 replied:
let us say there is a default rt, and default has dev associated it, and there must be a single IP  on that device, otherwise why do u know which one is default?

This code is not invented by me, please see thread for various discussion.  https://stackoverflow.com/questions/23558425/how-do-i-get-the-local-ip-address-in-go.

To me this is most precise solution of the bunch.

suwang48404 replied:
U are correct that if there are multiple interfaces on the node, the IP picked by default route may not be best to use to communicate with some peer nodes;  That will perhaps be another discussion how to handle multiple IPs on the worker node? 

regarding to node agent having route to internet. Yes, the underlying assumption is that worker node should have a default route. can u think of scenario where worker node does not require default route?

suwang48404 replied:
we now uses k8s ip to retrieve interface IP/mask

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -174,12 +174,24 @@ func (c *client) InstallNodeFlows(hostname string,
+		return nil
+	}
+
+	flows := []binding.Flow{
```
not sure it is a good idea to mix preallocation with append,

i.e.

   test := make([]string, 2, 3)
    test = append(test, "two", "three")
    fmt.Printf("%#v", test)
===> ]string{"", "", "two", "three"}

yes we can do assignment by index when preallocate, does not seem to me a good idea to scarifies code clarity for this supposed tiny optimization gain,

suwang48404 commented on cmd/antrea-agent/config.go:
```diff
@@ -65,4 +65,10 @@ type AgentConfig struct {
+	// Encap(default): inter-node Pod traffic is always encapsulated and Pod to outbound traffic is masqueraded.
+	// NoEncap: inter-node Pod traffic is not encapsulated, but Pod to outbound traffic is masqueraded
+	// Hybrid: noEncap if worker Nodes on same subnet, otherwise encap.
+	// PassThrough: antrea-agent provides network services at pod ingress/egress. Network connectivity is delegated to underlying network.
```
done

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -92,6 +120,7 @@ func (c *Client) SetupRules() error {
 		// Accept external-to-Pod traffic. This allows NodePort traffic to be forwarded even if the default FORWARD policy is DROP.
 		{FilterTable, AntreaForwardChain, []string{"!", "-i", c.hostGateway, "-o", c.hostGateway}, AcceptTarget, nil, "Antrea: accept external to pod traffic"},
 		// Mark Pod-to-external traffic which are received via host gateway interface but not sent via it for later masquerading in NAT table.
+		// TBD, qualifies it with peer CIDRs in no encap mode.
```
done. Removed all reference to passthrough

suwang48404 commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -469,21 +465,12 @@ func (c *Controller) addNodeRoute(nodeName string, node *v1.Node) error {
-	// present (as is the case for agent restarts).
-	if err = netlink.RouteReplace(route); err != nil {
-		return fmt.Errorf("failed to install route to Node %s with netlink: %v", nodeName, err)
+	routes, err := c.routeClient.AddPeerCIDRRoute(peerPodCIDR, c.gatewayLink.Attrs().Index, peerNodeIP, peerGatewayIP)
```
done. Removed all references to passthrough

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -118,6 +156,126 @@ func (c *Client) SetupRules() error {
+// It's idempotent and can be safely called on every startup.
+func (c *Client) AddPeerCIDR(peerPodCIDR *net.IPNet, peerNodeIP net.IP) error {
+	if !c.nodeConfig.TrafficEncapMode.NeedEncap(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+		ruleSpec := []string{
```
done

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -260,10 +286,23 @@ func (c *client) initialize() error {
+
+	if c.nodeConfig.TrafficEncapMode.SupportsNoEncap() {
+		if err := c.flowOperations.Add(c.l2ForwardOutputToRemoteInPortOut(types.HostGatewayOFPort, cookie.Default)); err != nil {
+			return fmt.Errorf("failed to install l2 forward in port out flow: %v", err)
```
done

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,222 @@
+				Gw:        peerNodeIP,
+			})
+		}
+	}
```
I made some comment change, hopefully it is clearer.
In the nutshell,
in encap mode, only service traffic outbound from gw0, single main rt is required to route the service traffic after LB back to gw0
in noencap mode, both pod and service traffic outbound from gw0, just main rt is required to route pod traffic to uplink, and separate service rt is required to route service traffic after LB back to gw0.

suwang48404 commented on pkg/agent/util/net_unix.go:
```diff
@@ -0,0 +1,66 @@
+		}
+		if dev != nil {
+			break
+		}
```
good suggestion. done.

suwang48404 commented on pkg/agent/types/node_config.go:
```diff
@@ -27,12 +31,40 @@ const (
+	PodEncapMode   PodEncapMode
+	GatewayConfig  *GatewayConfig
+	ServiceCIDR    *net.IPNet
+	ServiceRtTable *ServiceRtTableConfig
```
serviceRtTable is now moved to route.go, and comment was added there.

suwang48404 commented on cmd/antrea-agent/options.go:
```diff
@@ -88,6 +90,9 @@ func (o *Options) validate(args []string) error {
 	if o.config.OVSDatapathType != ovsconfig.OVSDatapathSystem && o.config.OVSDatapathType != ovsconfig.OVSDatapathNetdev {
 		return fmt.Errorf("OVS datapath type %s is not supported", o.config.OVSDatapathType)
 	}
+	if ok, _ := types.GetPodEncapModeFromStr(o.config.PodTrafficEncapMode); !ok {
```
done

suwang48404 commented on pkg/agent/types/pod_encap_mode.go:
```diff
@@ -0,0 +1,73 @@
+	return m == PodEncapModeEncap || m == PodEncapModeHybrid
+}
+
+func (m PodEncapMode) UseTunnel(peerIP net.IP, localIP *net.IPNet) bool {
```
done

suwang48404 commented on test/integration/agent/route_test.go:
```diff
@@ -0,0 +1,234 @@
+
+func TestRouteTable(t *testing.T) {
+	if _, incontainer := os.LookupEnv("INCONTAINER"); !incontainer {
+		t.Logf("Skip test not run in container")
```
done

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,333 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is required for Antrea to work properly in a Kind cluster on Linux. It takes care of
```
done.

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,333 @@
+}
+
+_usage="
+Usage: $0 create CLUSTER_NAME] [--pod-cidr POD_CIDR] [--antrea-cni true|false ] [--num-workers NUM_WORKERS] [--images IMAGES] [--subnets SUBNETS]
```
done

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,333 @@
+  modify-node: modify kind node with name NODE_NAME
+  --pod-cidr: specifies pod cidr used in kind cluster, default is $POD_CIDR
+  --encap-mode: inter-node pod traffic encap mode, default is encap
+  --antrea-cni: specifies use antrea or default cni in kind cluster, default is $ANTREA_CNI
```
update the help string to reflect no cni installed when antrea-cni is false.

suwang48404 commented on docs/kind.md:
```diff
@@ -56,6 +56,14 @@ antrea-agent-zsztq                   2/2     Running   0          8m56s
+### Short Cut
+ Alternatively to create a two worker node cluster with antrea installed, do
+
+  ./ci/kind/kind-setup.sh create CLUSTER_NAME
```
done

suwang48404 commented on docs/kind.md:
```diff
@@ -56,6 +56,14 @@ antrea-agent-zsztq                   2/2     Running   0          8m56s
+
+  ./ci/kind/kind-setup.sh create CLUSTER_NAME
+
+  To see other available options, do
```
added some more explanation.

suwang48404 commented on hack/generate-manifest.sh:
```diff
@@ -23,6 +23,7 @@ function echoerr {
 _usage="Usage: $0 [--mode (dev|release)] [--kind] [--keep] [--help|-h]
 Generate a YAML manifest for Antrea using Kustomize and print it to stdout.
         --mode (dev|release)  Choose the configuration variant that you need (default is 'dev')
+        --encap-mode          Inter-node pod traffic encapitulation mode.
```
done.

suwang48404 commented on pkg/agent/agent.go:
```diff
@@ -410,11 +426,25 @@ func (i *Initializer) setupGatewayInterface() error {
 func (i *Initializer) setupDefaultTunnelInterface(tunnelPortName string) error {
 	tunnelIface, portExists := i.ifaceStore.GetInterface(tunnelPortName)
+
+	if !config.TrafficEncapMode.SupportsEncap() {
```
To me I allow the design principal on encapsulation, caller need not details of how tunneling is setup.

wenyingd replied:
Actually, tunnel is not needed in noEncap mode, it is possible to mislead readers to believe tunnel  is needed by all networking mode if they don't go deep into the implementation of  "setupDefaultTunnelInterface"

suwang48404 replied:
true. but I generally prefer to implement branching logic at the lowest level. In this particular case, I use this function to remove previously created tunnel (when mode change from encap --> noEncap). I can do the remove logic also outside, IMO that would be messier and harder to read.

suwang48404 commented on pkg/agent/types/node_config.go:
```diff
@@ -27,12 +28,23 @@ const (
+	Bridge         string
+	Name           string
+	PodCIDR        *net.IPNet
+	NodeIPAddr     *net.IPNet
```
NodeIPAddr containers both IP/Mask hence net.IPNet is used

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,302 @@
+
+import (
+	"fmt"
+	"golang.org/x/sys/unix"
```
done

suwang48404 commented:
@antoninbas @tnqn @wenyingd @jianjuns 
Hi guys, do u mind to give this one last round of review? if they looks good, can u provide +1? I have one outstanding issue with Jianjun regarding to whether to expose NoEncap mode.

thx, Su

wenyingd commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,302 @@
+		_ = c.RemoveServiceRouting()
+		return nil
+	}
+	f, err := os.OpenFile("/etc/iproute2/rt_tables", os.O_RDWR|os.O_APPEND, 0)
```
What about using a constant for file path `/etc/iproute2/rt_tables`?
Another question, does this file need to be mounted from the host?

suwang48404 replied:
this file requires no mounting, it exits in each network namespace.

suwang48404 replied:
used a constant for routeConfigPath. done

wenyingd commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -192,6 +195,30 @@ func (c *client) gatewayClassifierFlow(gatewayOFPort uint32, category cookie.Cat
+
+// patchPortClassifierFlow generates the flow to mark ARP traffic comes from the patch port and
+// are destined to local pods
+func (c *client) patchPortClassifierARPFlow(patchPort uint32, podIP net.IP) binding.Flow {
```
It looks you would like to let the peer host gateway reply ARP request in NoEncap mode, we have installed an ARP responder flow entry by default( in Encap mode), please think about if that flow is still needed.

suwang48404 replied:
I was think ahead for passthrough mode ... but I can remove them...

suwang48404 replied:
will keep them for subsequent changes, if u don't mind

wenyingd commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -279,6 +306,18 @@ func (c *client) l2ForwardOutputFlow(category cookie.Category) binding.Flow {
+func (c *client) l2ForwardOutputToRemoteInPortOut(gwPort uint32, category cookie.Category) binding.Flow {
+	return c.pipeline[l2ForwardingOutTable].BuildFlow(priorityHigh).MatchProtocol(binding.ProtocolIP).
+		MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
+		MatchInPort(gwPort).MatchRegRange(int(portCacheReg), gwPort, ofPortRegRange).
```
There might be two kinds of traffic sent to the OVS bridge: 1) local pod to remote pod 2) local Pod to Service, and then to the backend Pod. For 1), this logic is correct, for we want to use host routing/iptables to forward the packet to the remote Node. For 2), if the backend Pod is also on the same node, this flow might send the packet to a wrong target.

suwang48404 replied:
This flows refers service traffic enters gw0 to LB and goes out again to a pod on a different node.
if pod backing a service is from local node, the load balanced pod-to-pod traffic will not hit this flow.

say backend pod is on the same node, service traffic will be

request: container(dstIP=serviceIP, srcIP=containerIP) ===> gw0(egress to host network, inport != outpor = gw0, this flow is not hit) ==> LB (dstIP=backEndlContainerIP, srcIP=serviceIP) ==> gw0(ingress from host network) ===> backend container via l2 flow 

Notice in this data flow the l2ForwardOutputToRemoteInPortOut flow was not hit?

wenyingd replied:
Got it, you also use the destinationed OpenFlow port(gw0) as one match condition in this flow, so it would not be hit by the packet targets at local containers. Thanks for explanation.

wenyingd commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -192,6 +195,30 @@ func (c *client) gatewayClassifierFlow(gatewayOFPort uint32, category cookie.Cat
+func (c *client) patchPortClassifierARPFlow(patchPort uint32, podIP net.IP) binding.Flow {
+	classifierTable := c.pipeline[classifierTable]
+	return classifierTable.BuildFlow(priorityNormal).
+		MatchInPort(patchPort).
```
It would be better to explicately add `MatchProtocol(binding.ProtocolARP)`

suwang48404 replied:
that is good point, will do.

suwang48404 replied:
added binding arp and ip protocol, done.

lzhecheng commented:
Please push your patch again to refresh jenkins jobs' status.

tnqn commented on pkg/agent/agent.go:
```diff
@@ -191,6 +202,8 @@ func (i *Initializer) initInterfaceStore() error {
 
 func (i *Initializer) Initialize() error {
 	klog.Info("Setting up node network")
+	config.TrafficEncapMode = i.trafficEncapMode
```
Is there a way to inject the mode to structs which need it instead of using a global variable which could break the e encapsulation?

suwang48404 replied:
It was in NodeConfig, JianJun suggested to make it global because it is technically not a node configuration. I am fine either way.

jianjuns replied:
@suwang48404 : sorry it might be my fault not to explain my thoughts clearly. I do not mean to save the parameter to a global variable (we definitely should not do that for OO design), but meant either make it a separate field of AgentInitializer (and pass it to other interfaces/funcs as a separate parameter too) or define a new struct for all global network parameters (encapMode, ipsecPSK, serviceCIDR, tunnelType, etc.) and make it a field of AgentInitializer (and pass to other interfaces/funcs).

suwang48404 replied:
MHO, passing same parameter to variable client/module is not good approach. If we have many parameters from agent-config.conf, it means we must pass each of them to a bunch of clients/modules, pretty messy.

as they are immutable, it'd be better to define them once, and be accessed freely by module/clients

my 2 cents, Su

suwang48404 replied:
That is good point and I agree. Let me open an issue and hope to address this holistically.
I hope that OK as the problem we are address preexists to this PR.

jianjuns replied:
@suwang48404 : I am glad to refactor it more later. However, for this PR, I would suggest at least to pass this separate parameter like other parameters, instead of using a global variable. The whole Antrea code has no global variable.

suwang48404 replied:
Ok, passed encapMode to individual consumers.

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -101,6 +132,19 @@ func (c *Client) SetupRules() error {
+			rule{table: MangleTable, chain: PreRoutingChain, parameters: nil, target: AntreaMangleChain, targetOptions: nil, comment: "Antrea: jump to Antrea mangle rule"},
+			//  Marks service traffic in PREROUTING chain so that service traffic may use ip rule to pick up the service route table.
+			rule{table: MangleTable, chain: AntreaMangleChain, parameters: []string{"-i", c.hostGateway, "-d", c.serviceCIDR.String()},
+				target: MarkTarget, targetOptions: []string{"--set-xmark", rtTblSelectorMark}, comment: "Andrea: mark service traffic"},
```
```suggestion
				target: MarkTarget, targetOptions: []string{"--set-xmark", rtTblSelectorMark}, comment: "Antrea: mark service traffic"},
```

suwang48404 replied:
thx for notice will do

suwang48404 replied:
done

tnqn commented on test/integration/agent/iptables_test.go:
```diff
@@ -38,47 +41,83 @@ func TestSetupRules(t *testing.T) {
+-A ANTREA-POSTROUTING -m mark --mark 0x400/0x400 -m comment --comment "Antrea: masquerade traffic requiring SNAT" -j MASQUERADE`},
+			{table: "mangle", output: `:ANTREA-MANGLE - [0:0]
+-A PREROUTING -m comment --comment "Antrea: jump to Antrea mangle rule" -j ANTREA-MANGLE
+-A ANTREA-MANGLE -d 1.1.0.0/16 -i gw0 -m comment --comment "Andrea: mark service traffic" -j MARK --set-xmark 0x800/0x800`},
```
```suggestion
-A ANTREA-MANGLE -d 1.1.0.0/16 -i gw0 -m comment --comment "Antrea: mark service traffic" -j MARK --set-xmark 0x800/0x800`},
```

suwang48404 replied:
done

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -118,6 +162,130 @@ func (c *Client) SetupRules() error {
+// It's idempotent and can be safely called on every startup.
+func (c *Client) AddPeerCIDR(peerPodCIDR *net.IPNet, peerNodeIP net.IP) error {
+	if !config.TrafficEncapMode.NeedEncap(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+		// The default masquerading rule is all traffic in-port=gw0, out-put!=gw0. The rule
```
```suggestion
		// The default masquerading rule is all traffic in-port=gw0, out-port!=gw0. The rule
```

suwang48404 replied:
done

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -118,6 +162,130 @@ func (c *Client) SetupRules() error {
 		if err := c.ensureRule(rule.table, rule.chain, ruleSpec); err != nil {
 			return err
 		}
+		c.storeRule(rule.table, rule.chain, ruleSpec)
```
why we need to store them in-memory? Didn't see the cache are consumed after initialization.

suwang48404 replied:
because iptables changes between different mode, we do reconciliation as agent start. 
The logic is as follows:

1. save all rules installed so far
2. after route_controller finishes process all the known nodes during its own reconcile stage, it calls iptable.Reconcile.
3. iptable.Reconcile retrieves all iptable entries that refers to and is in Antrea-xxx chain, if any of the rule does not match that in 1) is removed.

tnqn replied:
I see the problem now, it should still not work, as route_controller only remove stale routes and ports but not install required iptables, `iptable.Reconcile` will still remove nodes' rules, executing the Reconcile in route_controller has no difference with executing inside `iptables.SetupRules`.
And do we miss the logic to remove iptables rules for nodes in node deletion handling?

suwang48404 replied:
This is what happens with regards to iptable and route table reconciliation, hope it helps,

1. route.Reconcile logic removes all stale routes from last incarnation
2.iptable.Reconcile logic removes all stale iptable rules from last instance. Including those rules setup in iptables.SetupRules in previous instance, and are no longer needed now.
3. We don't have logic to iptable rule deletion when node delete because all CIDRs present must be part of cluster CIDR.

tnqn replied:
1. Between `iptables.SetupRules` and `iptables.Reconcile`, there are no other iptables rules (for node CIDRs) are installed,`iptables.Reconcile` will still remove previously installed and still required existing rules, and even if it's expected, there's no point to put `iptables.Reconcile` in route_controller as the cleanup can just be executed in `iptables.SetupRules`. Note that `removeStaleGatewayRoutes` and `removeStaleTunnelPorts` won't remove routes and ports that belong to existing nodes.
2. That would lead to orphaned iptables rule, though harmless, but the same applies to almost every resources including routes, openflows, container interfaces, aren't orphaned resources a bug?
@jianjuns @antoninbas

suwang48404 replied:
The problem was that iptable.reconcile may be called before nodes had the opportunity to add per podCIDR rules. 

The solution is that before calling iptable.reconcile, node_route_controller iterate through all known nodes, and add per podCIDR rule for each of them.

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -154,6 +322,89 @@ func (c *Client) ensureRule(table string, chain string, ruleSpec []string) error
+	}
+	return false
+}
+
```
Perhaps using iptables-restore operations like kube-proxy will make the processing simpler, but not needed to be in this PR of course.

suwang48404 replied:
I have not looked into Kube-proxy, not sure how was iptable-save, iptable-restore was used in their context. Of course there is always opportunity to improve.

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
@jianjuns 
would u like me to remove NoEncap mode all together from this PR until routed solution is ready? 
let me know.

suwang48404 commented:
> > would u like me to remove NoEncap mode all together from this PR until routed solution is ready?
> 
> Yes, If you think ok, I prefer to remove NoEncap for now, until we see real use cases. But like to hear other's opinions too.

would remove Antrea-agent.conf description and config.go only make sense to u? That way, the feature is not exposed? and can be easily added back if there is use case? meanwhile backend stays the same?

jianjuns commented:
I would keep the option in config.go and maybe antrea-agent.conf, if we keep the code. Otherwise we might forget the code? And we should test it in CI if we are able to, even we just test with a single subnet.

jianjuns commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -32,3 +32,10 @@
+# Determines how traffic is encapsulated. It has the following options
+# encap(default): Inter-node Pod traffic is always encapsulated and Pod to outbound traffic is masqueraded.
+# noEncapSameL2:  Inter-node Pod traffic is not encapsulated, but Pod to outbound traffic is masqueraded.
+#                 Worker nodes must be on the same L2 segment.
```
Nit: nodes -> Nodes. It is minor, but in most code/documentation, we use Node but not worker Node to follow K8s API, so maybe just use Node.

suwang48404 replied:
@jianjuns 

I think u r correct all alone. It just dawns on me that NoEncapSameL2 is just subnet of NoEncap(routed) because even if there are multiple subnets on the same l2 segment, there will be a gw interface somewhere for each of subnet.

Will make a quick change
1. move NoEncapSameL2 to NoEncap
2. in noEncap mode, for peer node on different subnet, forward to node gw.

Thx, Su

suwang48404 replied:
move back to noEncap node, the noEncap mode routing behavior now is
1. if peer node is on the same subnet, peer node is the next hop of pod CIDR
2. if not, node's route table is used to route pod CIDR assuming there is underlying route support.

suwang48404 commented:
/test-all

suwang48404 commented:
@jianjuns @tnqn @antoninbas @wenyingd 
hi all,

I appreciate you spending time review this large PR. I believe at this point I have addressed all ur suggestions/comments/concerns. If so, can one of u kindly provide approval and bring closure to it?

Thx, Su

tnqn commented on pkg/agent/agent.go:
```diff
@@ -215,6 +225,11 @@ func (i *Initializer) Initialize() error {
+		return err
+	}
+
+	klog.Infof("Agent initialized NodeConfig=%v\nencapMode=%s", i.nodeConfig, config.TrafficEncapMode)
```
nit: better not to introduce newline which could break log format.

suwang48404 replied:
done

tnqn commented on pkg/agent/config/traffic_encap_mode.go:
```diff
@@ -0,0 +1,83 @@
+	TrafficEncapModeEncap TrafficEncapModeType = iota
+	TrafficEncapModeNoEncap
+	TrafficEncapModeHybrid
+	TrafficEncapModeInvald = -1
```
```suggestion
	TrafficEncapModeInvalid = -1
```

suwang48404 replied:
done

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -42,24 +56,34 @@ var (
-	masqueradeMark  = fmt.Sprintf("%#08x/%#08x", masqueradeValue, masqueradeValue)
+	masqueradeMark  = fmt.Sprintf("%#x/%#x", masqueradeValue, masqueradeValue)
+
+	// RtTblSelectorValue selects which route table to use to forward service traffic back to host gateway gw0.
```
I think `RtTblSelectorValue` is not `which route table`? It should be the mark value by which service traffic is identified?

suwang48404 replied:
I think either description is OK? It is true this mark identifies the service traffic, but the propose of that identification is to select the right route table for service traffic.

traffic --> service traffic (iptable set RtTblSelectorValue) ----> ip rule matches RtTblSelectorValue and select service route table.

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -118,6 +162,130 @@ func (c *Client) SetupRules() error {
+			return nil
+		}
+		if err := c.ipt.Insert(FilterTable, AntreaForwardChain, 1, ruleSpec...); err != nil {
+			return fmt.Errorf("AddPeerCIDR  Insert %s: %w", ruleSpec, err)
```
```suggestion
			return fmt.Errorf("AddPeerCIDR Insert %s: %w", ruleSpec, err)
```

suwang48404 replied:
done

tnqn commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -279,6 +308,18 @@ func (c *client) l2ForwardOutputFlow(category cookie.Category) binding.Flow {
 
+// l2ForwardOutputFlow generates the flow that forward peer node traffic via gw0.
+// This flow supersedes default output flow because ovs by default auto-skips packets with output = input port.
+func (c *client) l2ForwardOutputToRemoteInPortOut(gwPort uint32, category cookie.Category) binding.Flow {
```
The name in comment is still `l2ForwardOutputFlow`.
The function name is a little hard to understand, how about `l2ForwardOutputInPortFlow` given it's a special case of `l2ForwardOutputFlow`.

suwang48404 replied:
done

tnqn commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -332,6 +373,21 @@ func (c *client) l3FwdFlowToRemote(
 		Done()
 }
 
+// l3FwdFlowToRemote generates the L3 forward flow on source node to support traffic to remote via gateway.
```
```suggestion
// l3FwdFlowToRemoteViaGw generates the L3 forward flow on source node to support traffic to remote via gateway.
```

suwang48404 replied:
done

tnqn commented:
> @jianjuns @tnqn @antoninbas @wenyingd
> hi all,
> 
> I appreciate you spending time review this large PR. I believe at this point I have addressed all ur suggestions/comments/concerns. If so, can one of u kindly provide approval and bring closure to it?
> 
> Thx, Su

Thanks @suwang48404 for your work and patience and sorry for my late review! Just came back from the long holiday and only available today to do a full review, I have added my comments, I think some of them might be risky, we can discuss more inline, I will prioritize reviewing of this PR in following days.

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(open), err %v", err)
+	}
+	defer func() { _ = f.Close() }()
```
my editor complains about unhandled return value, if I do defer f.close(). It is easier on my eyes

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+	AntreaServiceTable = "Antrea-service"
+	// AntreaServiceTableIdx is route table index for Antrea service traffic.
+	AntreaServiceTableIdx = 300
+	routeTableConfigPath  = "/etc/iproute2/rt_tables"
```
updating this file creates additional route table, its content is consumed by network stack. please see, for instance,
https://blog.scottlowe.org/2013/05/29/a-quick-introduction-to-linux-policy-routing/

tnqn replied:
I think the blog is a little misleading, a route table is not created by adding an entry to the file,  it just gives a name to a route table number while the number can be used directly without it. And in the implementation of netlink package and this PR, only number (`serviceRtTableConfig.idx`) is ever used, so I doubted why we need to have code to update the file.
You can verify it by executing commands manually:
```
root@dev4docker:~/antrea# ip rule add iif gw0 table 310
root@dev4docker:~/antrea# ip route add 1.1.1.1 dev gw0  table 310
root@dev4docker:~/antrea# ip route show table 310
1.1.1.1 dev gw0  scope link
root@dev4docker:~/antrea# ip rule
0:      from all lookup local
32756:  from all iif gw0 lookup 310
32766:  from all lookup main
32767:  from all lookup default
```

suwang48404 replied:
sure I can remove that routeTableConfigPath

suwang48404 replied:
per discussion, we decided keep the route table name for debugging purpose.

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+		if err := netlink.RouteReplace(route); err != nil {
+			klog.Fatalf("Failed to add link route to service table, err=%v", err)
+		}
+	}
```
kernel adds direct connect route to the main route table, this function add direct connect route to our additional service traffic route table.

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+	}
+
+	if !ServiceRtTable.IsMainTable() {
+		filter.Table = 0
```
if no tableIdx is given to netlink, it defaults the operation to the main route table. 
L196-L202 collect routes on gw0 from service route table
L204-210, if service table != main table, collect routes on gw0 from main route table.

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+		// rt is reference to actual data, as it changes,
+		// it cannot be used for assignment
+		tmpRt := rt
+		rtMap[rt.Dst.String()] = &tmpRt
```
route.go is agnostic to Node? Not complete sure what u have in mind?

suwang48404 replied:
change List return type to map[string][]*netlink.Route, and Delete to input argument to []*netlink.Route as requested.

suwang48404 commented on build/yamls/antrea-ipsec.yml:
```diff
@@ -264,6 +264,13 @@ data:
+
+    # Determines how traffic is encapsulated. It has the following options
+    # encap(default): Inter-node Pod traffic is always encapsulated and Pod to outbound traffic is masqueraded.
+    # noEncapSameL2:  Inter-node Pod traffic is not encapsulated, but Pod to outbound traffic is masqueraded.
```
done

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+	oldTablesRaw := make([]byte, 1024)
+	bLen, err := f.Read(oldTablesRaw)
+	if err != nil {
+		klog.Fatalf("Unable to create service route table(read), err %v", err)
```
done

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+		LinkIndex: util.GetNetLink(c.nodeConfig.GatewayConfig.Link).Attrs().Index}
+	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, filter, netlink.RT_FILTER_TABLE|netlink.RT_FILTER_OIF)
+	if err != nil {
+		return fmt.Errorf("route table(flush): %w", err)
```
done.

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -101,6 +132,19 @@ func (c *Client) SetupRules() error {
 
+	if config.TrafficEncapMode.SupportsNoEncap() {
+		rules = append(rules,
+			// 	Creates AntreaManagle chain in mangle table PREROUTING chain.
```
done

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -101,6 +132,19 @@ func (c *Client) SetupRules() error {
+			//  Marks service traffic in PREROUTING chain so that service traffic may use ip rule to pick up the service route table.
+			rule{table: MangleTable, chain: AntreaMangleChain, parameters: []string{"-i", c.hostGateway, "-d", c.serviceCIDR.String()},
+				target: MarkTarget, targetOptions: []string{"--set-xmark", rtTblSelectorMark}, comment: "Antrea: mark service traffic"},
+			// 	Creates AntreaRow chain in raw table PREROUTING chain.
```
done

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+	err = netlink.RuleAdd(ipRule)
+	if err != nil {
+		if os.IsExist(err) {
+			klog.Errorf("rule already created")
```
added priority value to ip rule to prevent redundant ip rule creation,

tnqn replied:
That doesn't stop duplicate rules from being created:
```
root@dev4docker:~/antrea# ip rule add iif gw0 lookup 310 priority 32756
root@dev4docker:~/antrea# ip rule add iif gw0 lookup 310 priority 32756
root@dev4docker:~/antrea# ip rule
0:      from all lookup local
32756:  from all iif gw0 lookup 310
32756:  from all iif gw0 lookup 310
```
It has to do a check before creating.

In case you miss another comment, the log should NOT be Error if the rule already exists as it could be a normal restart.

suwang48404 replied:
I think netlink priority maps to pref from ip rule command

root@f1c86e7fe988:/# ip rule add iif eth0 pref 300 table 300
root@f1c86e7fe988:/# ip rule add iif eth0 pref 300 table 300
RTNETLINK answers: File exists
root@f1c86e7fe988:/# echo $?
2

I did test with existing netlink implementation and verified it.

suwang48404 replied:
I am not sure why ur set up did not detect. netlink does not have api to check for rule existence,

suwang48404 replied:
thx for the pointer, used netlink.ruleList to detect redundancy.

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+func (c *Client) AddPeerCIDRRoute(peerPodCIDR *net.IPNet, gwLinkIdx int, peerNodeIP, peerGwIP net.IP) ([]*netlink.Route, error) {
+	if peerPodCIDR == nil {
+		klog.Errorf("Empty peer pod CIDR")
+		return nil, nil
```
done

suwang48404 commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,303 @@
+		}
+		retIdx++
+	}
+	return routes[:retIdx], err
```
retIdx is residual of previous changes, where I thought of return partial routes. Now the logic is either all routes are created successfully, or none at all. retIdx is removed,

suwang48404 commented on pkg/agent/types/node_config.go:
```diff
@@ -27,12 +28,23 @@ const (
+}
+
+func (n *NodeConfig) String() string {
+	return fmt.Sprintf("\nNodeName: %s\nPodCIDR: %s\nNodeIP: %s\nDev:%s\nGateway: %s\n",
```
done

suwang48404 commented:
> > @jianjuns @tnqn @antoninbas @wenyingd
> > hi all,
> > I appreciate you spending time review this large PR. I believe at this point I have addressed all ur suggestions/comments/concerns. If so, can one of u kindly provide approval and bring closure to it?
> > Thx, Su
> 
> Thanks @suwang48404 for your work and patience and sorry for my late review! Just came back from the long holiday and only available today to do a full review, I have added my comments, I think some of them might be risky, we can discuss more inline, I will prioritize reviewing of this PR in following days.


Hi @tnqn 

I appreciate ur detailed review, per our offline discussion, u have 4 main concerns,
1. global variable(encapMode)
2. iptables reconciling removes valid iptable rules
3. net_unix gets IP via different way from K8s
4, duplicate ip rules

This most recent change should have address all of them, along with some miscellaneous stuff. Please let me know if I have missed anything or that there are additional concerns

Thx, Su

suwang48404 commented:
/test-all

suwang48404 commented on pkg/agent/agent.go:
```diff
@@ -78,16 +83,19 @@ func NewInitializer(
 	tunnelType ovsconfig.TunnelType,
-	enableIPSecTunnel bool) *Initializer {
+	enableIPSecTunnel bool,
+	inEncapMode string,
```
"in" means input, renamed it to encapModeStr,

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,304 @@
+
+	gwConfig := c.nodeConfig.GatewayConfig
+	if gwConfig != nil && c.nodeConfig.PodCIDR != nil {
+		// add local podCIDR if applicable to service rt table
```
Nit: add -> Add. Add '.' at end of the line.

suwang48404 replied:
done

jianjuns commented on pkg/agent/util/net_unix.go:
```diff
@@ -0,0 +1,57 @@
+			}
+		}
+	}
+	return nil, nil, fmt.Errorf("unable to find local ip and device")
```
Nit: ip -> IP

suwang48404 replied:
done

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -0,0 +1,304 @@
+		} else if !c.encapMode.NeedsRoutingToPeer(peerNodeIP, c.nodeConfig.NodeIPAddr) {
+			routes = append(routes, &netlink.Route{
+				Dst:       peerPodCIDR,
+				LinkIndex: util.GetNetLink(c.nodeConfig.NodeDefaultDev).Attrs().Index,
```
Do we must provide the dev in the route? I thought dev is optional?

suwang48404 replied:
I am not sure, but do u see problem provide dev? It make clear to me where we want to remove the route from

jianjuns replied:
No, but if we could also avoid to look up the default interface/IP, but just use the node IP, it sounds better to me. I would not introduce another dependency, and would not discover host routes if possible.

jianjuns replied:
Revoke what I said since I found you need to look up the subnet anyway. But for this route, I feel better to let host stack decide which dev to use (for example there can be a more optimal route than using the Node IP dev)?

suwang48404 replied:
line 164-169 is for peer node in hybrid or noEncap on same subnet as local node, in another word,  localIP, peerIP in same subnet, and localIP is assigned on default device. Thus if system find a better outbound interface (then directly connected route), either system messed up or there is bug is this Antrea route implementation.
what do u think?

jianjuns replied:
Maybe the other NICs perform better or user just wants to use it for Pod traffic? The IP can move from one NIC to other too and we have no way to detect? But my point is if it is not necessary why we worry about which dev to use?

suwang48404 replied:
The question is, is NodeIP externally reachable  IP or not, i.e it is not 127.0.01.  The way Antrea and I suspect every other implementation assumes is that node IP is externally reachable IP. In that context, I think it is safe to say it does not matter which interface pod traffic goes out of for as long as interface is connected to the same physical l2 segment?

suwang48404 replied:
sounds good, will change

suwang48404 replied:
done

jianjuns commented on pkg/agent/agent.go:
```diff
@@ -448,12 +480,20 @@ func (i *Initializer) initNodeLocalConfig() error {
+	}
 
-	i.nodeConfig = &types.NodeConfig{Name: nodeName, PodCIDR: localSubnet}
+	i.nodeConfig = &types.NodeConfig{Name: nodeName, PodCIDR: localSubnet, NodeIPAddr: localAddr, NodeDefaultDev: dev.Attrs().Name}
```
Why not just use the IP on K8s Node? Anyway for remote Nodes we know only their IPs from K8s.

suwang48404 replied:
IP from k8s node is single IP, whereas NodeIPAddr is IP/Mask, needed to subnet testing with peer node

jianjuns replied:
I am wrong here. I got you need to know the subnet.

suwang48404 commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -192,6 +197,32 @@ func (c *client) gatewayClassifierFlow(gatewayOFPort uint32, category cookie.Cat
 
+// patchPortClassifierFlow generates the flow to mark IP traffic comes from the patch port and
+// are destined to local pods
+func (c *client) patchPortClassifierIPFlow(patchPort uint32, podIP net.IP) binding.Flow {
```
this is intended for policy only mode

suwang48404 commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -432,19 +461,22 @@ func (c *Controller) addNodeRoute(nodeName string, node *v1.Node) error {
-		// security policies.
-		if err = c.createIPSecTunnelPort(nodeName, peerNodeIP); err != nil {
-			return err
+	if c.encapMode.NeedsEncapToPeer(peerNodeIP, c.nodeConfig.NodeIPAddr) {
```
done

suwang48404 commented on pkg/agent/agent.go:
```diff
@@ -198,11 +210,8 @@ func (i *Initializer) Initialize() error {
-		return fmt.Errorf("error creating iptables client: %v", err)
-	}
-	if err := iptablesClient.SetupRules(); err != nil {
+	i.iptablesClient.Initialize(i.hostGateway, i.serviceCIDR, i.nodeConfig, i.trafficEncapMode)
```
merged SetupRules into Intialize.

suwang48404 commented on pkg/agent/iptables/iptables.go:
```diff
@@ -101,6 +134,19 @@ func (c *Client) SetupRules() error {
+				target: MarkTarget, targetOptions: []string{"--set-xmark", rtTblSelectorMark}, comment: "Antrea: mark service traffic"},
+			// 	Creates AntreaRaw chain in raw table PREROUTING chain.
+			rule{table: RawTable, chain: PreRoutingChain, parameters: nil, target: AntreaRawChain, targetOptions: nil, comment: "Antrea: jump to Antrea raw rule"},
+			//  Allows re-entering pod-to-pod traffic identified by src-mac=ReentraceMAC to bypass conntrack.
```
done

tnqn commented on pkg/agent/agent.go:
```diff
@@ -448,12 +479,20 @@ func (i *Initializer) initNodeLocalConfig() error {
 	return nil
 }
 
-// getNodeName returns the node's name used in Kubernetes, based on the priority:
```
the comment is removed by accident?

suwang48404 replied:
done

tnqn commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -269,6 +266,37 @@ func (c *Controller) removeStaleTunnelPorts() error {
 }
 
+func (c *Controller) reconcileIPTables() error {
+
```
this newline is not needed?

suwang48404 replied:
done

tnqn commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -269,6 +266,37 @@ func (c *Controller) removeStaleTunnelPorts() error {
+		}
+		err = c.iptablesClient.AddPeerCIDR(podCIDR, peerNodeIP)
+		if err != nil {
+			klog.Errorf("Failed to get update iptables for node %s  : %v", node.Name, err)
```
```suggestion
			klog.Errorf("Failed to update iptables for node %s: %v", node.Name, err)
```

suwang48404 replied:
done

tnqn commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -381,13 +409,12 @@ func (c *Controller) syncNodeRoute(nodeName string) error {
-		}
-		c.installedNodes.Store(nodeName, nil)
+	entry, flowsAreInstalled := c.installedNodes.Load(nodeName)
+	routes := entry.([]*netlink.Route)
```
Need to check `routes != nil` like before, otherwise the conversion would panic if the node is not installed before.

suwang48404 replied:
r u concerned about iterate through nil? go can iterate through nil slice, it produced 0 iterations. but compiles and runs.

func main() {
    var test []string 
    for _, t := range test {
        fmt.Printf("no harm, %s", t)
    }
    fmt.Print("done")
}

tnqn replied:
No, my concern is the conversion, nil (type interface{}) can not be converted to the slice. Try this: https://play.golang.org/p/AXHOGXPZvay

suwang48404 replied:
done.

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -118,6 +161,130 @@ func (c *Client) SetupRules() error {
+	for _, table := range []string{NATTable, FilterTable, MangleTable, RawTable} {
+		chains, err := c.ipt.ListChains(table)
+		if err != nil {
+			return fmt.Errorf("iptables list chain for %s : %w", table, err)
```
```suggestion
			return fmt.Errorf("iptables list chain for %s: %w", table, err)
```

suwang48404 replied:
done

tnqn commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -192,6 +197,32 @@ func (c *client) gatewayClassifierFlow(gatewayOFPort uint32, category cookie.Cat
 		Done()
 }
 
+// patchPortClassifierFlow generates the flow to mark IP traffic comes from the patch port and
```
```suggestion
// patchPortClassifierIPFlow generates the flow to mark IP traffic comes from the patch port and
```

suwang48404 replied:
done

tnqn commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -192,6 +197,32 @@ func (c *client) gatewayClassifierFlow(gatewayOFPort uint32, category cookie.Cat
+		Done()
+}
+
+// patchPortClassifierFlow generates the flow to mark ARP traffic comes from the patch port and
```
```suggestion
// patchPortClassifierARPFlow generates the flow to mark ARP traffic comes from the patch port and
```

suwang48404 replied:
done

tnqn commented on pkg/agent/util/net_unix.go:
```diff
@@ -0,0 +1,57 @@
+)
+
+// GetLocalNodeAddr return a local IP/mask that is on the path of default route.
+func GetDefaultLocalNodeAddr(localIP net.IP) (*net.IPNet, netlink.Link, error) {
```
function name and implementation don't match its comment.

suwang48404 replied:
done

suwang48404 commented:
/test-all

jianjuns commented on pkg/agent/types/node_config.go:
```diff
@@ -27,12 +28,23 @@ const (
+	Name           string
+	PodCIDR        *net.IPNet
+	NodeIPAddr     *net.IPNet
+	NodeDefaultDev string
```
Do we still need this NodeDefaultDev? If not could we remove it?

suwang48404 replied:
done

suwang48404 commented:
/test-all

antoninbas commented on pkg/agent/iptables/iptables.go:
```diff
@@ -42,24 +56,35 @@ var (
 }
 
+// key is chain, value is rules in the chain
+type chainRules map[string][][]string
```
Could this comment explain why the value is `[][]string` and not `[]string`?

suwang48404 replied:
done

antoninbas commented on test/e2e/basic_test.go:
```diff
@@ -287,12 +292,24 @@ func TestReconcileGatewayRoutesOnStartup(t *testing.T) {
 			return false, err
 		}
-		if len(routes) < clusterInfo.numNodes-1 {
+		expectedRtNum := clusterInfo.numNodes - 1
```
for readability, could you use a `minNumRoutes` and a `maxNumRoutes` variables, and compute them before the `PollImmediate` call based on the encap mode?

suwang48404 replied:
done.

antoninbas commented on test/e2e/basic_test.go:
```diff
@@ -302,16 +319,19 @@ func TestReconcileGatewayRoutesOnStartup(t *testing.T) {
 
-	routeToDelete := routes[0]
+	var routeToDelete *Route
+	if encapMode.SupportsEncap() {
```
Shouldn't we just skip this test altogether for `NoEncap`?
It may also be a good idea to update the comment for this test to explain the differences in the test for the different encap modes.

suwang48404 replied:
for noencap mode, test excises deleting stale CIDR routes.

suwang48404 commented on ci/kind/test-e2e-kind.sh:
```diff
@@ -0,0 +1,60 @@
+docker pull busybox
+docker pull nginx
+
+echo "======== Test encap mode =========="
```
Hi antonin, 

I has some reservations about sharing network resources, container resource to run all test cases at the same time. Tests may be interfering with each other or vm may running out of resources with three k8s clusters running.

I have a subsequent change ready for PR that uses a single test bed for all three mode, and each additional network mode only run a small set of network tests. This would also address the time run issue. Hope this will be OK with u.

resolve this for now, please feel free to re-open if this does not adequately address ur concern.

antoninbas replied:
I'm not asking for tests to be run on the same cluster. It is a very easy change to create 3 Github jobs that can run in parallel instead of running the tests sequentially in the same Github job. Similarly to how we run e2e tests in parallel with unit tests. Github does not impose a limit on the number of parallel jobs (unlike some other CI systems), it will let you use whichever resources are available.

suwang48404 replied:
got it.
break kind.yml into three cases one for each encap mode.

suwang48404 replied:
Antonin, 

what is expected test case? 

For instance, if I rename kind test to

"-name E2e encap tests on a Kind cluster on Linux"

the test would run, but there would be outstanding expectation, some thinkg

expect E2e tests on a Kind cluster on Linux ...., and as result tests requirements are not fulfilled.

change the kind test back to
"-name E2e tests on a Kind cluster on Linux" solve the problem.

I did not find anywhere in repro making reference to expected tests.

Thx, Su

antoninbas replied:
I believe you should keep everything inside kind.yml. The name of the "workflow"  is still "Kind", but you can have 3 entries in the "jobs" array:
 * E2e tests on a Kind cluster on Linux (encap)
 * E2e tests on a Kind cluster on Linux (hybrid)
 * E2e tests on a Kind cluster on Linux (noEncap)

These jobs should execute in parallel if Github as enough compute resources available.
But maybe I'm missing the question?

suwang48404 commented on Makefile:
```diff
@@ -28,16 +31,17 @@ bin:
 	$(error Cannot use target 'test-unit' on a non-Linux OS, but you can run unit tests with 'docker-test-unit')
 test-integration:
-	$(error Cannot use target 'test-integration' on a non-Linux OS)
+	$(error Cannot use target 'test-integration' on a non-Linux OS, but you can run unit tests with 'docker-test-integration')
```
done

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,330 @@
+
+# The script creates and deletes kind testbeds. Kind testbeds may be created with
+# docker images preloaded, antrea-cni preloaded, antrea-cni's encapsulation mode,
+# and docker bridge network conneting to worker node.
```
done

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,330 @@
+  --antrea-cni: specifies install Antrea CNI in kind cluster, default is true.
+  --num-workers: specifies number of worker nodes in kind cluster, default is $NUM_WORKERS
+  --images: specifies images loaded to kind cluster, default is $IMAGES
+  --subnets: a subnet creates a seperate docker bridge network with assigned subnet that worker nodes may connect to. Default is empty all worker
```
done.

suwang48404 commented on ci/kind/kind-setup.sh:
```diff
@@ -0,0 +1,330 @@
+  # force coredns to run on control-plane node because it
+  # is attached to docker0 bridge and uses host dns.
+  # Worker node may be configured to attach to custom bridges
+  # which uses dockerd as dns, causing coredns to detect
```
done

suwang48404 commented on ci/kind/test-e2e-kind.sh:
```diff
@@ -0,0 +1,60 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script is required for Antrea to work properly in a Kind cluster on Linux. It takes care of
```
done

antoninbas replied:
I don't think my comment was addressed. I simply don't believe the information about disabling "TX hardware checksum offload" is relevant here. `kind-setup.sh` is the script that does that.

suwang48404 commented on docs/kind.md:
````diff
@@ -56,6 +56,17 @@ antrea-agent-zsztq                   2/2     Running   0          8m56s
 antrea-controller-775f4d79f8-6tksp   1/1     Running   0          8m56s
 ```
 
+### Short Cut
````
done.

suwang48404 commented on pkg/agent/agent.go:
```diff
@@ -448,8 +482,17 @@ func (i *Initializer) initNodeLocalConfig() error {
+	}
+	localAddr, _, err := util.GetIPNetDeviceFromIP(ip)
+	if err != nil {
+		return fmt.Errorf("initNodeLocalConfig %v", err)
```
done.

suwang48404 commented on pkg/agent/config/traffic_encap_mode.go:
```diff
@@ -0,0 +1,81 @@
+	}
+)
+
+// GetTrafficEncapModeFromStr returns true and TrafficEncapModeType responding to input string.
```
done.

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

antoninbas commented on .github/workflows/kind.yml:
```diff
@@ -26,21 +28,52 @@ jobs:
+      - uses: actions/setup-go@v1
+        with:
+          go-version: 1.13
+      - name: Build Antrea image
```
There is probably a way to share the Antrea image between jobs, but we can leave that as an improvement for later.

suwang48404 replied:
will keep that in mind in next check in

thx,

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
/test-conformance

suwang48404 commented:
/test-all

suwang48404 commented:
/test-conformance

suwang48404 commented:
/test-conformance

suwang48404 commented:
/test-conformance
/test-networkpolicy

suwang48404 commented:
/test-conformance

suwang48404 commented:
/test-networkpolicy

suwang48404 commented:
@jianjuns can u provide approval now?
